docs/core/extensions/logger-message-generator.md

```diff
@@ -0,0 +1,320 @@
+
+As of .NET 6 Preview 4, we instroduced `LoggerMessageAttribute` to help auto-generated perfomant logging APIs using `Microsoft.Extensions.Logging` namespace. The soucrce generation logging support is designed to deliver a highly-usable and high-performance logging solution for modern .NET applications. We leverage the base ILogger model with `LoggerMessage.Define` exposed by .NET and augment it to simplify use and improve performance.
+
+The source generator gets triggered with a `LoggerMessageAttribute` on partial logging methods, and is able to either autogenerate the implementation of these partial methods, or produces compile-time diagnostics hinting to proper usage of this logging approach.
+
+The compile-time logging solution is typically considerably faster at runtime than existing logging approaches. It does this by eliminating boxing, temporary allocations, and copies to the maximum extent possible.
+
+Below is an example class with one declared logging method. The method is marked with an attribute and is declared static and partial. The code generator kicks in at build time and supplies an implementation of this partial method.
+
+```csharp
+static partial class Log
+{
+    [LoggerMessage(EventId = 0, Level = LogLevel.Critical, Message = "Could not open socket to `{hostName}`")]
+    public static partial void CouldNotOpenSocket(ILogger logger, string hostName);
+}
+```
+
+The above sample shows the canonical use of the logging source generator, where the logging method is static and the log level is specified in the attribute definition. There are other forms possible too. For example, in the following example the logging method is declared as an instance method. In this form, the logging method gets the logger by accessing an `ILogger` field in the containing class.
+
+```csharp
+class MyLogWrapper
+{
+    private readonly ILogger _logger;
+
+    public MyLogWrapper(ILogger logger)
+    {
+        _logger = logger;
+    }
+
+    [LoggerMessage(EventId = 0, Level = LogLevel.Critical, Message = "Could not open socket to `{hostName}`")]
+    public partial void CouldNotOpenSocket(string hostName);
+}
+
+```
+
+Sometimes, the logging level needs to be a dynamic property rather than being statically built into the code. You can do this by omitting the logging level from the attribute and instead supplying it as an argument to the logging method:
+
+```csharp
+static partial class Log
+{
+    [LoggerMessage(EventId = 0, Message = "Could not open socket to `{hostName}`")]
+    public static partial void CouldNotOpenSocket(ILogger logger, LogLevel level, string hostName);
+}
+```
+
+You can omit the logging message and an empty string will be provided for the message but the state will contain the arguments and formats them as key value pairs.
+
+```csharp
+using Microsoft.Extensions.Logging;
+using System.Text.Json;
+
+namespace LoggingGeneratorSampleApp
+{
+    class Program
+    {
+        static void Main(string[] args)
```

Comment:
Should C#'s top level statements be used here?
To reduce the noise and make the sample more concise?

Reply:
Done